### PR TITLE
feat: Forward operation in `StackLink.request()` when offline

### PR DESCRIPTION
**Pull request by Ldoppea**
> [!NOTE]
> This PR follows #1483 and #1486

On the Flagship app we want to serve `.query()` request using the `StackLink` when the device is connected, but we want to fallback to the `PouchLink` when we detect a connection loss

To allow this, we allow the consuming app to provide an `isOnline()` method to the `StackLink`

When provided, the `StackLink` will check for connectivity before doing its request. When offline, instead of processing the request, it will instead forward the request to the next `Link`

The code is generic so the consuming app can configure a `PouchLink` or anything else as the next `Link`

We want to also catch network errors when the `isOnline()` methods fails to detect connection loss, then we also fallback to the next `Link`

___

### Related PRs:
- #1483
- #1486
- cozy/cozy-flagship-app#1209

### CODE REVIEW

File: packages/cozy-client/src/StackLink.js [hunk 1]
```diff
@@ -77,11 +80,26 @@ export default class StackLink extends CozyLink {
...
-      return this.executeMutation(operation, result, forward)
+  async request(operation, result, forward) {
+    if (this.isOnline && !(await this.isOnline())) {
+      return forward(operation)
     }
-    return this.executeQuery(operation)
+
+    try {
+      if (operation.mutationType) {
+        return await this.executeMutation(operation, result, forward)
```
Comment (paultranvan) [inline packages/cozy-client/src/StackLink.js:90]: Why did you add the `await`?

File: packages/cozy-client/src/StackLink.js [hunk 2]
```diff
@@ -77,11 +80,26 @@ export default class StackLink extends CozyLink {
...
+      }
+      return await this.executeQuery(operation)
+    } catch (err) {
+      if (isReactNativeOfflineError(err)) {
+        return forward(operation)
```
Comment (paultranvan) [inline packages/cozy-client/src/StackLink.js:95]: I assume this is a fallback in case the `this.isOnline()` was not defined or wrongly returned true? I see the value, but I'm wondering whether it is good to have this kind of logic here. 
What if someone has a Cozy react-native app with only the StackLink defined? Here, it will forward the operation and the app will never receive the error

File: packages/cozy-client/src/StackLink.js (hunk 1 above)
Comment (Ldoppea) [reply packages/cozy-client/src/StackLink.js:90]: Without the await, the `try...catch` would not be able to catch anything executed inside the `executeMutation()` method.

This is because we directly return the Promise, so the `try...catch` will directly release the flow control without waiting for the `executeMutation` result.

If we add an `await` then the `try...catch` will continue its action until the Promise is fulfilled.

File: packages/cozy-client/src/StackLink.js (hunk 2 above)
Comment (Ldoppea) [reply packages/cozy-client/src/StackLink.js:95]: > I assume this is a fallback in case the this.isOnline() was not defined or wrongly returned true?

Yes, this is mainly because when a device loose the Internet connection, it can take a few seconds before the offline status is triggered. This often happens when you are connected to an Internet box through Wifi. If you unplug the internet connection, then your computer/phones will still displayed the "online" icon for a few seconds.

> What if someone has a Cozy react-native app with only the StackLink defined? Here, it will forward the operation and the app will never receive the error

Good catch. But I'm not sure that we can check if any `next link` exist. I'll investigate how to do this correctly.

File: packages/cozy-client/src/StackLink.js (hunk 1 above)
Comment (paultranvan) [reply packages/cozy-client/src/StackLink.js:90]: Very clear, thank you

File: packages/cozy-client/src/utils.js [hunk 3]
```diff
@@ -68,3 +68,15 @@ export const hasQueriesBeenLoaded = queriesResults => {
...
+/**
+ * Check is the error is about ReactNative not having access to internet
+ *
+ * @param {Error} err - The error to check
+ * @returns {boolean} True if the error is a network error, otherwise false
+ */
+export const isReactNativeOfflineError = err => {
+  // This error message is specific to ReactNative
+  // Network errors on a browser would produce another error.message
+  return err.message === 'Network request failed'
```
Comment (paultranvan) [inline packages/cozy-client/src/utils.js:81]: I wonder how stable this message is, and if it could change in future RN upgrades...

**Review by paultranvan — APPROVED** [review]
Looks good, but I have a concern on error handling when only StackLink is used
